### ISSUE TITLE
Refactor train_curriculum.py - Split 525-line training script into modules

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/bucket-brigade#80]
## Summary

The `scripts/train_curriculum.py` file has grown to **525 lines**. While it's a cohesive training script, it contains multiple distinct responsibilities that could be separated for better maintainability and reusability.

## Current Structure Analysis

The file contains:

1. **Neural Network Architecture** (~50 lines):
   - `PolicyNetwork` class with shared layers and action heads
   - Forward pass and action sampling logic

2. **Training Utilities** (~50 lines):
   - `compute_gae()` - Generalized Advantage Estimation
   - `evaluate_policy()` - Policy evaluation

3. **PPO Training Logic** (~120 lines):
   - `train_stage_ppo()` - Single-stage PPO training
   - Trajectory collection
   - Loss computation and optimization
   - TensorBoard logging

4. **Curriculum Management** (~180 lines):
   - `CurriculumTrainer` class
   - Curriculum definition (5 stages)
   - Multi-stage training orchestration
   - Checkpoint management
   - Cross-scenario evaluation

5. **CLI & Main** (~90 lines):
   - Argument parsing (12 arguments)
   - Seed setting
   - Entry point

6. **Imports & Docs** (~35 lines)

## Proposed Refactoring

Split into focused modules:

```
scripts/
├── train_curriculum.py          # Main entry point (reduced to ~100 lines)
└── bucket_brigade/
    └── training/
        ├── __init__.py
        ├── networks.py          # PolicyNetwork, value/policy heads
        ├── ppo.py               # PPO algorithm (GAE, train_step, etc.)
        ├── curriculum.py        # CurriculumTrainer, stage definitions
        ├── evaluation.py        # Policy evaluation utilities
        └── config.py            # Training configuration dataclass
```

### Detailed Module Breakdown

**`networks.py`** (~80 lines):
- `PolicyNetwork` class
- Network architecture helpers
- Action sampling utilities

**`ppo.py`** (~100 lines):
- `compute_gae()`
- `compute_ppo_loss()`
- `train_step()` - single update step
- Trajectory collection helpers

**`curriculum.py`** (~150 lines):
- `CurriculumStage` dataclass
- `CurriculumTrainer` class
- Default curriculum definitions
- Multi-stage training orchestration

**`evaluation.py`** (~50 lines):
- `evaluate_policy()`
- Cross-scenario evaluation
- Metrics collection utilities

**`config.py`** (~50 lines):
- `TrainingConfig` dataclass
- Hyperparameter definitions
- Argument parsing utilities

**`train_curriculum.py`** (main, ~100 lines):
- Import training components
- Parse arguments
- Create config
- Run training
- Save results

## Benefits

- **Reusability**: PPO/GAE code can be used for other training scripts
- **Testability**: Each module can be unit tested independently
- **Configurability**: Easier to experiment with different network architectures
- **Maintainability**: Changes to PPO algorithm isolated from curriculum logic
- **Documentation**: Each module has focused, clear purpose
- **Extension**: Easy to add new training algorithms alongside PPO

## Alternative: Lighter Refactoring

If full module split is too heavy, consider:

1. **Extract PolicyNetwork** to `bucket_brigade/models/policy_network.py`
2. **Extract CurriculumTrainer** to `bucket_brigade/training/curriculum.py`
3. Keep PPO helpers in training script (they're tightly coupled)

This gives 80% of the benefits with 40% of the effort.

## Acceptance Criteria

- [ ] Training script produces identical results (given same seed)
- [ ] All hyperparameters configurable via CLI
- [ ] Checkpoints backward compatible
- [ ] TensorBoard logging preserved
- [ ] Each new module under 200 lines
- [ ] Module imports clean and well-documented

## References

- Current file: `scripts/train_curriculum.py:525`
- Related: `bucket_brigade/envs/puffer_env.py` (training interface)
- Related: Issue #79 (engine.rs refactoring)

## Notes

This refactoring would also make it easier to:
- Add new RL algorithms (A2C, SAC, etc.)
- Experiment with different network architectures
- Implement distributed training
- Add model ensembles or distillation

**Comment by rjwalters** [comment]
📋 **Guide Priority Assessment**

**Hermit Proposal**: Refactor train_curriculum.py (525 lines → modular structure)

**Assessment**: Good proposal for code maintainability, but NOT urgent.

**Considerations**:
- **Impact**: Improves code reusability for RL training, enables future work
- **Risk**: Low-Medium - training scripts are less critical than core engine
- **Effort**: 1 day of refactoring + testing to ensure identical results
- **Dependencies**: Could benefit from resolving #10/#11 first (training work)

**Current Project Context**:
- Training script just added in PR #75 (curriculum learning)
- TensorBoard logging added in PRs #61, #74
- Training issues #10, #11 currently blocked
- Recent focus on RL training features

**Recommendation**:
- ✅ Approve for work eventually - enables better RL experimentation
- ⏸️ NOT urgent - training script works fine as-is
- 🔄 Consider after #10/#11 resolve - might inform refactoring needs
- 🎯 Priority: LOW-MEDIUM (technical debt, but RL-focused)

**Alternative**: The "lighter refactoring" option mentioned in the issue (just extract PolicyNetwork and CurriculumTrainer) might be better - 80% benefit for 40% effort.

**Suggested User Action**: 
- Consider lighter refactoring option first
- Or approve with "loom:issue" but as lower priority than unblocking #10
- Builder should focus on training work (#10, #11) before code organization

**Comment by rjwalters** [comment]
## 📋 Curator Enhancement

This issue is well-scoped and the proposed refactoring makes sense. I've analyzed the actual script (550 lines, not 525) and the existing codebase structure to provide detailed implementation guidance.

---

## Implementation Guidance

### 1. Recommended Approach: **Lighter Refactoring** (as suggested in issue)

Given that:
- The training script was recently added (PR #75)
- There are existing blocked training issues (#10, #11)
- The script is working correctly as-is

I recommend the **lighter refactoring option** first (80% benefit, 40% effort):

#### Phase 1: Extract Core Components (Recommended)
```
bucket_brigade/
└── training/
    ├── __init__.py           # Clean exports
    ├── networks.py           # PolicyNetwork class (~80 lines)
    └── curriculum.py         # CurriculumTrainer class (~200 lines)

scripts/
└── train_curriculum.py      # Reduced to ~250 lines (PPO logic + CLI)
```

**Rationale**: Keep the tightly-coupled PPO training logic in the script since it's specific to this curriculum approach. Extract only the truly reusable components.

#### Phase 2: Full Module Split (Optional, after #10/#11 resolve)
If additional training scripts emerge (A2C, SAC, distributed training), then proceed with full module split as originally proposed.

### 2. Module Breakdown Details

#### `bucket_brigade/training/networks.py` (~80 lines)
```python
"""Policy network architectures for reinforcement learning."""

import torch
import torch.nn as nn
from typing import List, Tuple, Optional

class PolicyNetwork(nn.Module):
    """
    Actor-critic policy network with multi-discrete action space.
    
    Architecture:
    - Shared feature extraction layers
    - Separate heads for each action dimension
    - Value estimation head
    
    Args:
        obs_dim: Observation space dimension
        action_dims: List of action space sizes for each dimension
        hidden_size: Hidden layer size (default: 64)
    
    Example:
        >>> policy = PolicyNetwork(obs_dim=42, action_dims=[10, 2], hidden_size=128)
        >>> action_logits, value = policy(obs_batch)
    """
    # [Current implementation from train_curriculum.py lines 28-78]
```

**Key points**:
- Keep exact same interface for backward compatibility
- Add comprehensive docstrings
- Add type hints for better IDE support
- Consider adding `from_checkpoint()` class method for loading

#### `bucket_brigade/training/curriculum.py` (~200 lines)
```python
"""Curriculum learning orchestration."""

import torch
from dataclasses import dataclass
from typing import Dict, List, Any
from pathlib import Path

@dataclass
class CurriculumStage:
    """Configuration for a single curriculum stage."""
    name: str
    num_opponents: int
    steps: int
    description: str
    progression_threshold: float

class CurriculumTrainer:
    """
    Manages progressive training through scenario curriculum.
    
    The curriculum starts with simple cooperation scenarios and
    progressively introduces more complex strategic challenges.
    
    Example:
        >>> trainer = CurriculumTrainer(args)
        >>> trainer.train_curriculum()
    """
    # [Current implementation from train_curriculum.py lines 257-458]
```

**Key points**:
- Use dataclass for `CurriculumStage` (cleaner than dict)
- Keep curriculum definition as class attribute
- Extract checkpoint saving/loading to separate methods
- Make curriculum externally configurable (JSON/YAML)

#### `bucket_brigade/training/__init__.py`
```python
"""Reinforcement learning training utilities."""

from .networks import PolicyNetwork
from .curriculum import CurriculumTrainer, CurriculumStage

__all__ = [
    "PolicyNetwork",
    "CurriculumTrainer", 
    "CurriculumStage",
]
```

### 3. Migration Strategy

**Step 1**: Create new module structure
```bash
mkdir -p bucket_brigade/training
touch bucket_brigade/training/__init__.py
touch bucket_brigade/training/networks.py
touch bucket_brigade/training/curriculum.py
```

**Step 2**: Move `PolicyNetwork` → `networks.py`
- Copy class with full implementation
- Add comprehensive docstrings
- Add type hints
- Keep interface identical

**Step 3**: Move `CurriculumTrainer` → `curriculum.py`
- Copy class with full implementation
- Import PPO training functions from parent script (for now)
- Add type hints and documentation

**Step 4**: Update `scripts/train_curriculum.py`
```python
from bucket_brigade.training import PolicyNetwork, CurriculumTrainer

# Keep PPO functions here for now:
# - compute_gae()
# - evaluate_policy()
# - train_stage_ppo()
```

**Step 5**: Verify identical behavior
```bash
# Run training with same seed, verify checkpoints are identical
python scripts/train_curriculum.py --seed 42 --run-name refactor_test

# Compare with baseline (if you saved one before refactoring)
```

### 4. Testing Strategy

#### Unit Tests
Create `tests/test_training.py`:
```python
import torch
import pytest
from bucket_brigade.training import PolicyNetwork, CurriculumTrainer

def test_policy_network_forward_pass():
    """Test PolicyNetwork produces expected shapes."""
    policy = PolicyNetwork(obs_dim=42, action_dims=[10, 2], hidden_size=64)
    obs = torch.randn(16, 42)  # Batch of 16
    
    action_logits, value = policy(obs)
    
    assert len(action_logits) == 2
    assert action_logits[0].shape == (16, 10)
    assert action_logits[1].shape == (16, 2)
    assert value.shape == (16, 1)

def test_policy_network_get_action_deterministic():
    """Test deterministic action selection."""
    policy = PolicyNetwork(obs_dim=42, action_dims=[10, 2], hidden_size=64)
    obs = torch.randn(1, 42)
    
    actions, log_prob, value = policy.get_action(obs, deterministic=True)
    
    assert len(actions) == 2
    assert log_prob is None  # Deterministic mode
    assert value.shape == (1, 1)

def test_policy_network_checkpoint_save_load():
    """Test checkpoint saving and loading preserves behavior."""
    policy1 = PolicyNetwork(obs_dim=42, action_dims=[10, 2], hidden_size=64)
    obs = torch.randn(1, 42)
    
    # Save
    checkpoint = {
        'policy_state_dict': policy1.state_dict(),
        'obs_dim': 42,
        'action_dims': [10, 2],
        'hidden_size': 64,
    }
    
    # Load
    policy2 = PolicyNetwork(
        checkpoint['obs_dim'], 
        checkpoint['action_dims'],
        checkpoint['hidden_size']
    )
    policy2.load_state_dict(checkpoint['policy_state_dict'])
    
    # Verify identical outputs
    with torch.no_grad():
        out1 = policy1(obs)
        out2 = policy2(obs)
        
    assert torch.allclose(out1[0][0], out2[0][0])
    assert torch.allclose(out1[1], out2[1])
```

#### Integration Test
```python
def test_curriculum_training_deterministic():
    """Verify curriculum training is deterministic with same seed."""
    # This is the critical test - ensures refactoring didn't break training
    # Run with minimal steps to keep test fast
    pass  # Implement minimal curriculum run
```

### 5. Backward Compatibility

**Critical**: Ensure existing checkpoints can still be loaded:

```python
# Old checkpoint format (from current script)
checkpoint = torch.load("models/old_run/curriculum_final.pt")

# Should still work after refactoring
from bucket_brigade.training import PolicyNetwork

policy = PolicyNetwork(
    checkpoint['obs_dim'],
    checkpoint['action_dims'], 
    checkpoint['hidden_size']
)
policy.load_state_dict(checkpoint['policy_state_dict'])
```

**Test this explicitly** to avoid breaking existing trained models.

### 6. Documentation Updates

Update documentation after refactoring:

1. **README.md** or **docs/training.md**: Document the training module structure
2. **Docstrings**: Add comprehensive docstrings to all extracted classes
3. **Type hints**: Add type hints for better IDE support
4. **Examples**: Add usage examples in docstrings

---

## Acceptance Criteria (Updated)

**Phase 1 (Lighter Refactoring)**:
- [ ] Create `bucket_brigade/training/` module
- [ ] Extract `PolicyNetwork` → `training/networks.py`
- [ ] Extract `CurriculumTrainer` → `training/curriculum.py`
- [ ] Update `scripts/train_curriculum.py` to import from new module
- [ ] Add comprehensive docstrings and type hints
- [ ] Verify identical training behavior (same seed → same results)
- [ ] All existing checkpoints load successfully
- [ ] Unit tests for `PolicyNetwork` (forward pass, action sampling, save/load)
- [ ] CI tests pass
- [ ] `scripts/train_curriculum.py` reduced to ~250 lines

**Phase 2 (Optional - Full Module Split)**:
- [ ] Extract PPO utilities → `training/ppo.py` (`compute_gae`, `train_step`)
- [ ] Extract evaluation → `training/evaluation.py`
- [ ] Extract config → `training/config.py` (dataclasses for hyperparameters)
- [ ] Multiple training scripts can reuse components
- [ ] Documentation updated with architecture diagram

---

## Test Plan

### Pre-Refactoring Baseline
1. Run full curriculum training with seed 42
2. Save final checkpoint and TensorBoard logs
3. Record final evaluation scores for all scenarios

### Post-Refactoring Validation
1. **Determinism Test**: Run with same seed, verify identical results
   ```bash
   python scripts/train_curriculum.py --seed 42 --run-name after_refactor
   ```
2. **Checkpoint Compatibility**: Load old checkpoints with new code
3. **Unit Tests**: Run all new unit tests for extracted modules
4. **Integration Test**: Verify full training run completes successfully
5. **Performance Test**: Ensure no performance regression (runtime within 5%)

### Acceptance
- [ ] All tests pass
- [ ] Training results match baseline (within numerical precision)
- [ ] Checkpoints backward compatible
- [ ] Code coverage ≥ 80% for new training module
- [ ] Documentation complete

---

## Implementation Effort Estimate

**Phase 1 (Lighter Refactoring)**:
- Module creation: 1 hour
- Extract and document classes: 2 hours
- Update imports and verify: 1 hour
- Write unit tests: 2 hours
- Integration testing: 1 hour
- Documentation: 1 hour

**Total: ~8 hours (1 day)**

**Phase 2 (Full Split)** - if needed later:
- Additional module extraction: 3 hours
- Testing and validation: 2 hours
- Documentation updates: 1 hour

**Total: +6 hours**

---

## Related Issues & PRs

- **Blocked by**: None
- **Blocks**: Future training work (will make it easier)
- **Related**: 
  - Issue #10: Training work (currently blocked) - might inform refactoring needs
  - Issue #11: Training work (currently blocked)
  - PR #75: Original curriculum learning implementation
  - Issue #79: Similar refactoring for `engine.rs`

---

## Recommendations

1. ✅ **Approve for work** - This is good technical debt reduction
2. ⏸️ **Priority: LOW-MEDIUM** - Not urgent, training works fine as-is
3. 🎯 **Start with Phase 1** (lighter refactoring) - Lower risk, faster delivery
4. 🔄 **Consider Phase 2 later** - Only if additional training scripts emerge
5. 💡 **Alternative**: Wait until #10/#11 resolve to see if they inform refactoring needs

---

## Notes

- The script is actually **550 lines** (not 525 as originally stated), but this doesn't change the refactoring approach
- The existing `evolution/` module provides a good template for how to structure the `training/` module
- Backward compatibility with existing checkpoints is critical - don't break trained models
- Consider making curriculum definition externally configurable (JSON/YAML) in Phase 2
